microsoft-edge/webview2/how-to/winrt-from-js.md

```diff
@@ -0,0 +1,416 @@
+
+1. Close the UWP sample app.  The DevTools window also closes.
+
+
+<!-- ====================================================================== -->
+## Make AddHostObjectToScript act like Chakra WinRT projection
+
+The `AddHostObjectToScript` feature defaults to using asynchronous and verbose proxies.  You can read more about that in the `AddHostObjectToScript` documentation; use the [See also](#see-also) links below.  To make `AddHostObjectToScript` act more like the Chakra WinRT projection, set the following properties:
+
+* `chrome.webview.hostObjects.option.defaultSyncProxy` - Proxies may either be asynchronous or synchronous.  Normally we know, when calling a method on a synchronous proxy, that the result should also be a synchronous proxy.  But in some cases, we lose that context, such as when providing a reference to a function to native code, and then native code later calling that function.  In these cases, the proxy will be asynchronous, unless this property is set.
+
+* `chrome.webview.hostObjects.options.forceAsyncMethodMatches` - This is an array of regular expressions.  If you call a method on a synchronous proxy, the method call will actually be performed asynchronously if the method name matches a string or regular expression that's in this array.  Setting this value to [/Async$/] will make any method that ends with `Async` be an asynchronous method call.  If an async method doesn't match here and isn't forced to be asynchronous, the method will be invoked synchronously, blocking execution of the calling JavaScript and then returning the resolution of the promise, rather than returning a promise.
+
+* `chrome.webview.hostObjects.options.ignoreMemberNotFoundError` - If you attempt to get the value of a property of a proxy, and the property doesn't exist on the corresponding native class, you'll get an exception - unless you set this property to `true`, in which case the behavior will match Chakra WinRT projection behavior (and general JavaScript behavior) and return `undefined` with no error.
+
+Chakra WinRT projection puts the WinRT namespaces directly on the root object.  In contrast:
+*  `AddHostObjectToScript` places async root proxies on `chrome.webview.hostObjects`.
+*  `AddHostObjectToScript` places sync root proxies on `chrome.webview.hostObjects.sync`.
+To access root proxies where Chakra WinRT projection code would expect, you can assign the root proxy WinRT namespace locations over to the root object. For example:
+
+```javascript
+window.Windows = chrome.webview.hostObjects.sync.Windows;
+```
+
+To ensure that the JavaScript that sets all this up is executed before anything else, you can either add the above statement to your JavaScript, or you can tell WebView2 to inject the above statement for you before running any other script, by using the `CoreWebView2.AddScriptToExecuteOnDocumentCreatedAsync` method.
+
+The following example demonstrates the above techniques:
+
+```javascript
```

Comment:
This code block is C++ not javascript. If we want to make it C# it would be the following (also removing the Cortana references)

```c#
webview.CoreWebView2.AddScriptToExecuteOnDocumentCreatedAsync(
            "(() => {" +
                    "if (chrome && chrome.webview) {" +
                        "console.log('Setting up WinRT projection options');" +
                        "chrome.webview.hostObjects.options.defaultSyncProxy = true;" +
                        "chrome.webview.hostObjects.options.forceAsyncMethodMatches = [/Async$/,/AsyncWithSpeller$/];" + 
                        "chrome.webview.hostObjects.options.ignoreMemberNotFoundError = true;"  + 
                        "window.Windows = chrome.webview.hostObjects.sync.Windows;" +
                    "}" +
                "})();");
```